Add subscriptions with lots of refactoring and code improvements

jfrerich opened the pull request:
## Summary

This PR was initially adding subscriptions but quickly turned into a refactor in order to share code among the different portions of the code base. If any reviewer prefers a 1:1 walkthrough of the functionality or code, please reach out!

### Add subscriptions (channel_header and post_menu locations)
Users can now add and update subscriptions through a modal

### Helper classes and functions
These were created in an attempt to provide importable packages for development in the future.  (still WIP, but a start)

Creating the following helpers allowed to clean up the folder structure to the app.  Now each of the following directories only contain implementations represented by the folder name

`bindings/` - only contains bindings
`call_handlers/` - only contains complex call_handlers. Simple call responses are in `restapi/fxxx` files. 
`forms/` - only contains forms

`src/utils/fields/app_fields.ts` - build new AppField objects
`src/utils/fields/fields_build.ts`- build `AppForm` `field`.  Array of visible AppFields in a modal
`src/utils/base_call_handler.ts` - call handler for handling lookup, form changes, and submissions
`src/utils/bindings.ts` - help funcs to build different location bindings
`src/utils/call_responses.ts` - helpers to build different `AppCallResponse` types

## TODO
- [x] fields_build.ts - remove AppCall portion into a `AppParser` helper class.  Instantiate in build and allow using in other files that don't need the builder methods https://mattermost.atlassian.net/browse/MM-33387
- [x] Add more options for subscriptions (now only supports `changed` condition) https://mattermost.atlassian.net/browse/MM-33384
- [x] Fix all TypeScript errors and warnings https://mattermost.atlassian.net/browse/MM-33385
- [x] push Forked version of node-zendesk https://mattermost.atlassian.net/browse/MM-33386
- [x] Create PR against node-zendesk https://mattermost.atlassian.net/browse/MM-33386
- [ ] Remove TODO items from code and create as issues
- [x] Use existing forms to populate the slash command https://mattermost.atlassian.net/browse/MM-33388
- [x] Add ability to delete a sub
- [x] Delete button needs to be RED https://mattermost.atlassian.net/browse/MM-33389

## Ticket Links

https://mattermost.atlassian.net/browse/MM-30038 - Subscription - discovery/spec - How to create in zendesk?
https://mattermost.atlassian.net/browse/MM-30603 - Subscripti0ns - implementation - design modal
https://mattermost.atlassian.net/browse/MM-32405 - Create mm-client package for posting as bot, user, or admin
https://mattermost.atlassian.net/browse/MM-32719 - Create `FieldsBuilder` helper class to create form fields array
https://mattermost.atlassian.net/browse/MM-32721 - Create `call_responses` utils helpers
https://mattermost.atlassian.net/browse/MM-32718 - Create `app_fields` helper class to create AppsFields
https://mattermost.atlassian.net/browse/MM-32722 - Create `Bindings` helper class

jfrerich commented:
@mickmister, when are you added to the MM org again, can you self request a review :)

catalintomai commented on src/utils/utils.ts:
```diff
@@ -1,24 +1,55 @@
 export const makeFormOption = (option: zdFormFieldOption): AppSelectOption => ({label: option.name, value: option.id.toString()});
 export const makeFormOptions = (options: zdFormFieldOption[]): AppSelectOption[] => options.map(makeFormOption);
 
+export const makeSubscriptionOption = (option: zdSubsriptionFieldOption): AppSelectOption => ({label: getDisplaySubTitleOption(option), value: option.id.toString()});
```
type: `zdSubsriptionFieldOption`

jfrerich replied:
done.

catalintomai commented on src/app/app.ts:
```diff
@@ -1,71 +1,125 @@
+    values: AppCallValues
 
+    constructor(call: AppCall) {
+        this.values = call.values;
```
nit: maintain the declaration ordering above (lines 25-26)

jfrerich replied:
done.

catalintomai commented on src/app/app.ts:
```diff
@@ -1,71 +1,125 @@
         // create the ticket in Zendesk
-        const ticket = await tryPromiseWithMessage(zdClient.tickets.create(zdTicket), 'Failed to create Zendesk ticket');
+        const createReq = zdClient.tickets.create(zdTicket);
+        const ticket = await tryPromiseWithMessage(createReq, 'Failed to create Zendesk ticket');
```
maybe rename this to `zdTicket` and rename the original `zdTicket` to smth. else? (to keep in sync with the `zdUser`, etc.)

jfrerich replied:
good nit catch. Done.

catalintomai commented on src/app/triggerFromForm.ts:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    // getJSONDataFields constructs the object text string for a trigger
+    getJSONDataFields(): string {
```
nit: add `trigger` to the function name

jfrerich replied:
I debated on the function names in this file before. Filename, interface, factory function, and class name all reference `Trigger` in their names.  I think every non-exported function could actually go the other way and be minimalized by removing all references to a trigger.

For example changing all of the following: 
getTriggerActions -> getActions()
getTriggerTitle -> getTitle()
getTriggerDescription -> getDescription()
getTriggerConditions -> getConditions()

Thoughts? Seems personal preferences differ in this regard, but I do believe consistency is important

catalintomai replied:
The reason I mentioned `getJSONDataFields` is with ine other exception all the other methods have Trigger in their name. I guess to maintain consistency we could use this prefix in all of them or in none (excepting the constructor?)

jfrerich replied:
done. I removed trigger from the names, minus the exceptions you mentioned.

catalintomai commented on src/bindings/channel_header.ts:
```diff
@@ -0,0 +1,34 @@
+
+import {ZDIcon, Routes, newChannelHeaderBindings, Bindings} from '../utils';
+
+// getChannelHeaderBindings returns the users slash command bindings
```
why "slash command"?

jfrerich replied:
I just missed changing the comment while probably copying from the slash_command.ts file.  Fixed.

catalintomai commented on src/restapi/fIncomingWebhooks.ts:
```diff
@@ -0,0 +1,61 @@
+
+    const adminClient = newMMClient().asAdmin();
+
+    // const botClient = newMMClient().asBot();
```
cleanup?

jfrerich replied:
yep. Done.

catalintomai commented on src/forms/create_ticket.ts:
```diff
@@ -20,3 +31,199 @@ export async function newCreateTicketForm(call: AppCall): Promise<AppForm> {
+    // when form is submitted
+    getMappedName(field: Users.Fields.UserField): string {
+        const strFieldType = String(field.type);
+        switch (true) {
```
why not a sequence of `if...else` s?

jfrerich replied:
done. I was expecting more cases, but did not turn out that way.

catalintomai commented on src/utils/helper_classes/fields/app_fields.ts:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    // isTextArea the text field will be multilined
+    isTextArea(): void {
```
would rename this to smth. else other than `is` (the `is`-prefixed ones have a `true/false` semantics)

jfrerich replied:
I removed this helper class completely and just construct an AppField where needed.

catalintomai commented on src/forms/create_ticket.ts:
```diff
@@ -20,3 +31,199 @@ export async function newCreateTicketForm(call: AppCall): Promise<AppForm> {
+
+    // addFormsToSelectField maps zendesk forms to a Mattermost select field
+    addFormsSelectField(): void {
+        const field = new StaticSelectField(CreateTicketFields.NameFormsSelect, makeFormOptions(this.zdTicketForms));
```
I'd rather that we have smth. like : `const field = new StaticSelectField(...) { label:Form; refresh:...; is_required:...}` -  less verbose and contained and remove the one line helper functions (we don't get any special benefit from using them as wrappers around the field setting - like extra validation, checks, etc.)

jfrerich replied:
I understand.  I suppose these "helpers" aren't necessary and I need to make a parallel ticket to add the types for each field, instead of implementing this helper.

jfrerich replied:
done.  I removed this class and just construct an AppField

larkox reviewed (changes requested):
Great work Jason. A few things here and there. I have a few files still to check, will take a look at another moment.

jfrerich commented on src/app/app.ts:
```diff
@@ -1,71 +1,132 @@
         // create the ticket in Zendesk
-        const ticket = await tryPromiseWithMessage(zdClient.tickets.create(zdTicket), 'Failed to create Zendesk ticket');
+        const createReq = zdClient.tickets.create(zdTicketPayload);
+        const zdTicket = await tryPromiseWithMessage(createReq, 'Failed to create Zendesk ticket');
```
What do you think is the correct course of action if this happens? As is, the `tryPromiseWithMessage` function will throw an error and the message will be shown in the modal as an error.
![image](https://user-images.githubusercontent.com/7575921/108091510-65eabd80-7041-11eb-9134-726b95201ba7.png)

larkox replied:
Maybe it was not clear to me that this throws an error and that it is catch somewhere else. If that is the behaviour, I am fine with it, but the code does not feel straight forward.

jfrerich replied:
also added to the error handling Jira ticket. I was trying to have a helper function that would to some error handling instead of having try/catch blocks all over the place. If this isn't clear, I'll adjust it and push for feedback in the error handling PR.

jfrerich commented on src/app/triggerFromForm.ts:
```diff
@@ -0,0 +1,124 @@
+            const subID = this.values[SubscriptionFields.SubSelect_Name].value;
+            this.addTriggerField('id', subID);
+        }
+        this.addTriggerField('title', this.getTriggerTitle());
```
👍 I'll change each to an `addXYZ` and call the `addTriggerField` inside each `addXYZ` function

jfrerich commented on src/app/triggerFromForm.ts:
```diff
@@ -0,0 +1,124 @@
+
+    isNewTrigger(): boolean {
+        const subID = this.values[SubscriptionFields.SubSelect_Name].value;
+        return Boolean(subID === SubscriptionFields.NewSub_OptionValue);
```
it's unnecessary. changed.

jfrerich commented on src/bindings/post_menu.ts:
```diff
@@ -1,36 +1,54 @@
+            call: {
+                url: Routes.App.BindingPathOpenCreateTicketForm,
+                expand: {
+                    post: 'all',
```
True.  Thanks.  I also realized that I was asking for expanded context when some calls don't required any expanded context.  I've removed the expand field from where not needed.

jfrerich commented on src/app/triggerFromForm.ts:
```diff
@@ -0,0 +1,124 @@
+
+export function newTriggerFromForm(values: AppCallValues, context: AppContext): [any, FieldValidationErrors] {
+    const trigger = new TriggerFromForm(values, context).getTrigger();
+    return [trigger, trigger.fieldValidationErrors];
```
I'm not sure that there will be custom field validation for fields.  If so, I can steal similar code from the create_ticket code.  I'll remove from here for now.

jfrerich replied:
done.

jfrerich commented on src/app/triggerFromForm.ts:
```diff
@@ -0,0 +1,124 @@
+        const pairs = TriggerFields.ActionValuePairs;
+        pairs[TriggerFields.ChannelIDKey] = channelID;
+
+        const objStrings = Object.entries(TriggerFields.ActionValuePairs).map((pair) => {
```
Of course this is what I should have done instead of constructing myself. And I should have searched for a package that deals with JSON data. 

Done.

jfrerich commented on src/call_handlers/subscriptions.ts:
```diff
@@ -0,0 +1,33 @@
+    // show subcription options
+    handleSubmit = async (): Promise<AppCallResponse> => {
+        const app = newApp(this.call);
+        return app.createZDSubscription();
```
@larkox, note in the latest commit.  `handleSubmit` now returns the promised `app.createZDSubscrption()` call.  If you look at that function, I am either returning the error call response (shows in the modal) or an OK Call response (ephemeral message)

jfrerich commented on src/clients/zendesk.ts:
```diff
@@ -1,14 +1,61 @@
+    const mmUserID = context.acting_user_id || '';
+    const token = oauthStore.getToken(mmUserID);
+    if (!token) {
+        throw new Error('Failed to get user access_token');
```
We have a ticket to discuss error handling. I don't have a great solution right now though.  I will create a ticket specific to Zendesk if that is ok.

jfrerich replied:
Ticket created. https://mattermost.atlassian.net/browse/MM-33555

jfrerich commented on src/app/app.ts:
```diff
@@ -1,71 +1,132 @@
-        const botToken = configStore.getBotAccessToken();
-        const botClient = newMMClient(botToken);
+        const addToChannelReq = adminClient.addToChannel(botUserID, this.context.channel_id);
+        await tryPromiseWithMessage(addToChannelReq, 'Failed to add bot to team');
```
Good question.  I don't have that answer right now and have created a ticket to investigate.  https://mattermost.atlassian.net/browse/MM-33556

jfrerich commented on src/app/triggerFromForm.ts:
```diff
@@ -0,0 +1,124 @@
+
+    getChannelIDobject(): {} {
+        const fieldName = SubscriptionFields.ChannelPickerSelect_Name;
+        const channelID = this.values[fieldName].value;
```
ticket created. https://mattermost.atlassian.net/browse/MM-33557

jfrerich commented:
@catalintomai, @larkox you can review the changes I have made.  Note that some of the questions, I have just created Jira tickets to be investigated and answered later. I would like to get this merged asap and start moving on to webhooks

larkox commented:
@jfrerich There are conflicts with the master branch. Do you mind solving them before reviewing?

jfrerich commented:
@larkox, I resolved the conflicts in the latest commit and finally have this PR merged with the latest `master` changes. It is also working with latest `master` branch of `mattermost-plugin-apps`

jfrerich commented:
@catalintomai, get I get another review please?  I have latest master merge working with the latest `mattermost-plugin-apps` master branch too!

catalintomai commented:
> @catalintomai, get I get another review please? I have latest master merge working with the latest `mattermost-plugin-apps` master branch too!

will do.

catalintomai commented on src/app/triggerFromForm.ts:
```diff
@@ -0,0 +1,115 @@
+    context: AppContext;
+    trigger: any
+
+    constructor(values: AppCallValues, context: AppContext) {
```
nit: have context as first arg.

jfrerich replied:
done.

catalintomai commented on src/app/triggerFromForm.ts:
```diff
@@ -0,0 +1,115 @@
+        }
+
+        // do not let subscriptions without a condition be created...
+        if (conditions.any.length === 0) {
```
should this be moved immediately after the `for`?

jfrerich replied:
I actually removed the code between this if block and the for block.  I don't recall the need for that code and think it's invalid

catalintomai commented on src/forms/create_ticket.ts:
```diff
@@ -20,3 +31,206 @@ export async function newCreateTicketForm(call: AppCall): Promise<AppForm> {
+    }
+
+    // setState sets state for the current Zendesk Form
+    async setState(): Promise<void> {
```
I would avoid using this naming, usually we are using `setState` to initialize the fields of React's `this.state`, which we do not use here (maybe we should?)

catalintomai commented on src/utils/call_responses.ts:
```diff
@@ -0,0 +1,47 @@
+    };
+}
+
+export function newErrorCallResponseWithMessage(error: string): AppCallResponse {
```
nit: rename `error` to `message` (to distinguish from the following `newErrorCallResponseWithFieldErrors`)

jfrerich replied:
fair enough. I'll explicitly use the `error` field with value `message`

catalintomai commented on src/utils/helper_classes/fields/fields_builder.ts:
```diff
@@ -0,0 +1,151 @@
+    }
+
+    // isFieldValueSelected checks if the AppField has a defined value
+    isFieldValueSelected(field: AppField): boolean {
```
do we need it?

catalintomai commented on src/utils/helper_classes/fields/fields_builder.ts:
```diff
@@ -0,0 +1,151 @@
+    }
+
+    // getFields returns an array of currently mapped fields saved in the mapper
+    getFields(): AppField[] {
```
do we need it?

catalintomai commented on src/utils/helper_classes/fields/fields_builder.ts:
```diff
@@ -0,0 +1,151 @@
+    }
+
+    // addFields adds multiple fields to the current fields list array.
+    addFields(fields: AppField[]): void{
```
is it being used?

catalintomai commented on src/utils/helper_classes/fields/fields_builder.ts:
```diff
@@ -0,0 +1,151 @@
+    setDefaultMinLength(value: number): void;
+}
+
+export function newFieldsBuilder(call: AppCall): Ibuilder {
```
do we need it?

jfrerich replied:
I do use these in the code base and feel it was a point of confusion for me early on.  Propose creating a ticket to discuss and address in a separate PR later.

jfrerich replied:
@catalintomai, Jira issue created https://mattermost.atlassian.net/browse/MM-33638

catalintomai reviewed (approved):
few small comments, LGTM.

mickmister reviewed (changes requested):
Overall looks good, very exciting 🙂 

I have quite a few requests and a few comments

jfrerich commented on README.md:
````diff
@@ -167,7 +172,7 @@ exports.rudderAnalytics = rudderAnalytics;
 ```
 
-### 3. Log message received and bindings are not received by the 
+### 3. Log message received and bindings are not received by the
````
yes it is.

jfrerich replied:
fixed.

jfrerich commented on README.md:
```diff
@@ -167,7 +172,7 @@ exports.rudderAnalytics = rudderAnalytics;
 var fetch_etag_1 = require("node-fetch");
```
done.

jfrerich commented on src/call_handlers/create_ticket.ts:
```diff
@@ -0,0 +1,44 @@
+    // submit the ticket
+    handleSubmit = async (): Promise<AppCallResponse> => {
+        try {
+            const app = newApp(this.call);
```
I get this error, so I reverted back to lower case

jfrerich replied:
![image](https://user-images.githubusercontent.com/7575921/110533252-637d0000-80e3-11eb-9164-a58a082ec75e.png)

jfrerich commented on src/bindings/slash_commands.ts:
```diff
@@ -1,45 +1,76 @@
+    getBindings = (): AppBinding[] => {
+        const bindings: AppBinding[] = [];
+        if (this.isConnected()) {
+            bindings.push(this.cmdDisconnect());
+            if (this.isSysadmin()) {
+                bindings.push(this.cmdSubscribe());
+            }
```
I'm fine with this.  Do we have a convention?  Alphabetically?

jfrerich replied:
done.

jfrerich commented on src/clients/zendesk.ts:
```diff
@@ -1,14 +1,61 @@
+    create(trigger: any): any;
+    search(query: string): any;
+}
+interface OauthTokens {
```
done.

jfrerich commented on src/utils/helper_classes/fields/fields_builder.ts:
```diff
@@ -0,0 +1,151 @@
+    // select field, it is the value field in the returned object.
+    getFieldLabelByName(fieldName: string): string {
+        if (this.values && this.values.values[fieldName]) {
+            // this is select field return the value field in the object
+            if (this.values.values[fieldName].label) {
+                return this.values.values[fieldName].label;
+            }
+
+            // this is any field type that doesn't have an option type value
+            return this.values.values[fieldName];
```
Agree with removing the comments.  done.

jfrerich commented on src/forms/subscriptions.ts:
```diff
@@ -0,0 +1,401 @@
+
+    getSubTriggerByID(subID: string): any {
+        const triggers = this.getChannelTriggers(this.getSelectedChannelID());
+        for (const trigger of triggers) {
+            if (String(trigger.id) === subID) {
+                return trigger;
+            }
+        }
```
👍 done.

jfrerich commented on src/restapi/fHelp.ts:
```diff
@@ -0,0 +1,45 @@
+
+export async function fHelp(req: Request, res: Response): Promise<void> {
+    let helpText = getHelpText();
+    helpText += getSysadminCommands(req.body.context.acting_user_id);
```
done.

jfrerich commented on src/restapi/fIncomingWebhooks.ts:
```diff
@@ -0,0 +1,59 @@
+    const zdClient = newZDClient(fakeBotContext);
+    const auditReq = zdClient.tickets.exportAudit(ticketID);
+    const ticketAudits = await tryPromiseWithMessage(auditReq, `Failed to get ticket audits for ticket ${ticketID}`);
+    const ticketAudit = ticketAudits.slice(-1)[0];
```
done.

jfrerich replied:
I prefer the pop statement.

jfrerich commented on src/utils/utils.ts:
```diff
@@ -1,24 +1,55 @@
 }
 
-type zdFormFieldOption = {
+export type zdFormFieldOption = {
```
I'm open to either and don't really have a preference, at the moment

jfrerich replied:
I'll capitalize them based on the suggestion.

jfrerich commented on src/utils/constants.ts:
```diff
@@ -21,15 +24,19 @@ const AppPaths = {
 
     // Call routes are callable routes, but not bound to a location
     CallPathSubmitOrUpdateCreateTicketForm: '/submit-or-update-create-ticket-form',
+    CallPathSubmitOrUpdateSubcriptionForm: '/submit-or-update-subcription-form',
```
I misspelled this all over the place. Thanks for catching

jfrerich replied:
Fixed all occurrences across the code base.

jfrerich commented on src/forms/create_ticket.ts:
```diff
@@ -20,3 +31,206 @@ export async function newCreateTicketForm(call: AppCall): Promise<AppForm> {
+    private getViewableFields(ticketFields: Users.Fields.UserField[], formIDs: number[]): Tickets.Field[] {
+        const fields = [];
+        ticketFields.forEach((field: Users.Fields.UserField) => {
+            // omit fields that do not show up in the create ticket modeal in Zendesk
```
done

jfrerich commented on src/utils/constants.ts:
```diff
@@ -76,14 +86,77 @@ export const ZDFieldTypes = {
+    SaveButtonLabel: 'Save',
+
+    NewSub_OptionValue: 'newsubscription',
+    NewSub_OptionLabel: 'New Subscription',
```
Agree that this should be consistent. Done.

jfrerich commented on src/forms/create_ticket.ts:
```diff
@@ -20,3 +31,206 @@ export async function newCreateTicketForm(call: AppCall): Promise<AppForm> {
+        ticketFields.forEach((field: Users.Fields.UserField) => {
+            // omit fields that do not show up in the create ticket modeal in Zendesk
+            // but are returned in the ticketFields query
+            const omitFields = ['Group', 'Status'];
```
Sure.  I just moved this variable declaration to the top of the file

jfrerich commented on src/utils/helper_classes/fields/fields_builder.ts:
```diff
@@ -0,0 +1,151 @@
+    }
+
+    // getFieldLabelByName returns the defined AppCall label for a field or an empty string
+    // this is a helper function for a caller to check the value of an
+    // AppField. The return value is the value of the field. In the case of a
+    // select field, it is the value field in the returned object.
```
Yes, this is way too much information. I drastically reduced it to the following: 
```
   // getFieldLabelByName returns the defined AppCall label for a field or an empty string
   // In the case of a select field, it is the label field in the returned object.

```

jfrerich commented on src/utils/bindings.ts:
```diff
@@ -0,0 +1,60 @@
+    constructor(userID: string) {
+        this.userID = userID;
+        this.isUserConnected = isUserConnected(userID);
+        this.isUserSysadmin = isUserSysadmin(userID);
```
yes, this got funky in the KVStore PR https://github.com/mattermost/mattermost-app-zendesk/pull/28

I actually removed the `isUserConnected()` and `isUserSysadmin()` from the construct and pass those in as booleans fields which are calculated beforehand.  

Here's the file form that PR. https://github.com/mattermost/mattermost-app-zendesk/blob/kvstore/src/utils/bindings.ts  

Maybe you could check that out and comment in that PR if needed?

jfrerich commented on src/bindings/channel_header.ts:
```diff
@@ -0,0 +1,31 @@
+
+// getChannelHeaderBindings returns the users command bindings
+export const getChannelHeaderBindings = (userID: string): AppBinding => {
+    return new ChannelHeaderBindings(userID).getBindings();
```
No, it didn't prove useful.  I removed this in the KV store PR. #28

jfrerich commented on src/bindings/index.ts:
```diff
@@ -1 +1,15 @@
-export {getBindings} from './bindings';
+import {AppBinding} from 'mattermost-redux/types/apps';
+
+import {Bindings} from '../utils';
```
I purposefully added bindings to `utils` because I thought we might make usage of it later.  I might end up stripping it altogether.

jfrerich commented on src/app/user.ts:
```diff
@@ -1,6 +1,12 @@
-export function isUserConnected(userID: string): string {
-    return oauthStore.getToken(userID);
+export function isUserConnected(userID: string): boolean {
+    return Boolean(oauthStore.getToken(userID));
```
Can we address this when you get to https://github.com/mattermost/mattermost-app-zendesk/pull/28?

jfrerich commented on README.md:
```diff
@@ -102,26 +110,23 @@ From [Zendesk Documentation:](https://developer.zendesk.com/rest_api/docs/suppor
 1. Fill in the following:
-    1. **Trigger Name:** Notify Mattermost App when ticket created
+    1. **Trigger Name:** `__mm_webhook__ channelID:<channelID> teamID:<teamID>`
+        1. The `__mm_webhook__` prefix will be used by the zendesk cloud app and is necessary
```
Yes.  Sorry, this was added in this PR, hadn't modified the README. I made some changes but will create a ticket to thoroughly scrub the README.

jfrerich replied:
https://mattermost.atlassian.net/browse/MM-33703

jfrerich commented on README.md:
```diff
@@ -178,3 +183,19 @@ Through system console -> enable oauth2 service provider
+### 4. Need a branched version of the node-zendesk client so that we can query
+webhooks by title
+
+In `node_modules/node-zendesk/lib/client/triggers.js` add the following lines
```
I plan on doing both of these things. I do hate having to add a new repo and keeping it around though. 

Ticket created.
https://mattermost.atlassian.net/browse/MM-33704

hanzei replied:
Open Source Friday?

jfrerich commented on src/forms/subscriptions.ts:
```diff
@@ -0,0 +1,401 @@
+
+        this.addSubCheckBoxes();
+
+        this.addSubmitButtons();
```
Agree 100% and removed the spaces. done.

jfrerich commented on manifest.ts:
```diff
@@ -0,0 +1,31 @@
+import {AppBindingLocations} from 'mattermost-redux/constants/apps';
+import {AppID} from './src/utils'
```
yes, this is a problem.  I've created a ticket https://mattermost.atlassian.net/browse/MM-33720

jfrerich commented on src/forms/subscriptions.ts:
```diff
@@ -0,0 +1,401 @@
+
+    // getBulletedList returns a bulleted list of items with options header
+    // pretext
+    getBulletedList(pretext: string, items: string[]): string {
```
done.

jfrerich commented on src/forms/subscriptions.ts:
```diff
@@ -0,0 +1,401 @@
+
+    // byName is a map filter function to retrieve a given fieldName from an
+    // array of conditions
+    byName(name: string): boolean {
```
It is a zendesk field.  The problem is that zendesk types aren't defined for many objects.  I added one for this case though because it's small.  also used locally, so I added it to this file.

jfrerich commented on src/app/app.ts:
```diff
@@ -1,71 +1,138 @@
+
+        // create the trigger object from the form response
+        let zdTriggerPayload: any;
+        try {
+            zdTriggerPayload = newTriggerFromForm(this.context, this.values);
+        } catch (e) {
+            return newErrorCallResponseWithMessage(e.message);
+        }
```
`tryPromiseWithMessage` throws a `new Error`.  In this case, if an error occurs, I what to send a call-response with the error message instead.

jfrerich replied:
put into catch-all NITs ticket mattermost.atlassian.net/browse/MM-33787

jfrerich commented on src/restapi/fDisconnect.ts:
```diff
@@ -10,26 +9,23 @@ import {oauthStore} from '../store';
     // from the zendesk app
-    const currentToken = await tryPromiseWithMessage(zdClient.oauthtokens.current(), 'failed to get current user token');
+    const oauthReq = zdClient.oauthtokens.current();
+    const currentToken = await tryPromiseWithMessage(oauthReq, 'failed to get current user token');
```
https://mattermost.atlassian.net/browse/MM-33744
and 
https://github.com/mattermost/mattermost-app-zendesk/issues/22

jfrerich commented on src/forms/subscriptions.ts:
```diff
@@ -0,0 +1,401 @@
+        }
+    }
+
+    getChannelTriggers(channelID: string): any[] {
```
Many of the types in Zendesk are not defined in the javascript client. added a ticket.

https://mattermost.atlassian.net/browse/MM-33745

jfrerich commented on src/app/triggerFromForm.ts:
```diff
@@ -0,0 +1,112 @@
+export class TriggerFromForm implements ITriggerFromFrom {
+    values: AppCallValues;
+    context: AppContext;
+    trigger: any
```
There wasn't one available for the client. I've added a ticket to replace occurrences of "any" throughout the code-base.

jfrerich commented on src/bindings/channel_header.ts:
```diff
@@ -0,0 +1,31 @@
+
+    channelHeaderSubscribe = (): AppBinding => {
+        return {
+            app_id: 'zendesk',
```
I'll fix this when it arrives :)

jfrerich commented on src/restapi/fIncomingWebhooks.ts:
```diff
@@ -0,0 +1,59 @@
+    // TODO: we need zendesk bot admin so that admin requests can be made by the
+    // bot and not from an actiing user
+    const fakeBotContext: AppContext = {
+        acting_user_id: 'rgixs6uimp88tq8x8w3yxu3oqe',
```
When a webhook is received, there is no information of and 'acting_user_id'.  This is hardcoded right now as my userID.  It is needed for creating posts and needs to be addressed. 
https://mattermost.atlassian.net/browse/MM-33749

jfrerich commented on src/utils/constants.ts:
```diff
@@ -76,14 +86,77 @@ export const ZDFieldTypes = {
+
+    // TODO this is the target id value. Need to store this value when creating the
+    // target through the app, or when the user has to create target in zendesk
+    TargetID: 360002371891,
```
Ticket created. I'll add to the config modal for others to use, but plan to create the trigger at install 
https://mattermost.atlassian.net/browse/MM-33750

jfrerich commented on src/forms/subscriptions.ts:
```diff
@@ -0,0 +1,401 @@
+            label: 'Optional message',
+            value: text,
+            readonly: true,
+        };
```
https://mattermost.atlassian.net/browse/MM-33762  ticket created and TODO removed.

jfrerich commented on src/forms/create_ticket.ts:
```diff
@@ -20,3 +31,206 @@ export async function newCreateTicketForm(call: AppCall): Promise<AppForm> {
+    }
+
+    isSystemField(field: Users.Fields.UserField): boolean {
+        return (SystemFields.includes(String(field.type)) || field.system_field_options);
```
ticket created to check for all String() castings.  https://mattermost.atlassian.net/browse/MM-33763

jfrerich commented on src/utils/helper_classes/fields/fields_builder.ts:
```diff
@@ -0,0 +1,151 @@
+        let value = f.value || this.getSavedValue(f.name);
+        if (f.type === AppFieldTypes.BOOL) {
+            value = true;
+            if (f.value === 'false') {
```
https://mattermost.atlassian.net/browse/MM-33763

jfrerich commented on src/app/app.ts:
```diff
@@ -1,71 +1,138 @@
             user_id: botUserID,
-            channel_id: String(context.channel_id),
-            root_id: String(context.post_id),
+            channel_id: String(this.context.channel_id),
```
ticket created to investigate all String() castings. https://mattermost.atlassian.net/browse/MM-33763

jfrerich commented on src/restapi/fIncomingWebhooks.ts:
```diff
@@ -0,0 +1,59 @@
+    const ticketLink = `[${ticketID}](${ZDHost}${ZDTicketPath}/${ticketID})`;
+
+    switch (auditEvent.type) {
+    case 'Comment':
```
ticket created https://mattermost.atlassian.net/browse/MM-33764

jfrerich commented on src/forms/subscriptions.ts:
```diff
@@ -0,0 +1,401 @@
+        if (this.isNewSub()) {
+            f.hint = this.getSelectedSubTriggerName();
+
+            // TODO this value needs to be set to empty so the hint will show.
+            // setting to '' does not work
+            // not setting keeps previous text form previous selected sub
+            f.value = this.getSelectedSubTriggerName();
```
I wasn't able to get this to work correctly and will investigate again.  I added a ticket. 
https://mattermost.atlassian.net/browse/MM-33765

jfrerich replied:
This works if you click New Subscription first.  However, if you click any other existing sub and then New Sub, the previous existing sub name remains as the text.  I cannot clear the text and have the hint show. I think that I will need to intercept the context and remove the value so it does not exist and get filled if a hint is needed.

jfrerich replied:
https://mattermost.atlassian.net/browse/MM-33765

hanzei commented on manifest.ts:
```diff
@@ -0,0 +1,32 @@
+
+type RequestedPermissions = 'act_as_user' | 'act_as_bot'
+
+export type Manifest = {
```
Why is this file needed given that we have `manifest.json`?

jfrerich replied:
put into catch-all NITs ticket https://mattermost.atlassian.net/browse/MM-33787

hanzei commented on src/app/app.ts:
```diff
@@ -1,71 +1,141 @@
+            return newErrorCallResponseWithMessage(e.message);
+        }
+
+        let request: any;
```
Can we type this?

jfrerich replied:
put into catch-all NITs ticket https://mattermost.atlassian.net/browse/MM-33787

hanzei commented on src/app/triggerFromForm.ts:
```diff
@@ -0,0 +1,112 @@
+
+    addConditions(): void {
+        const conditions = {
+            any: [],
```
Can we set the type here?

jfrerich replied:
put into catch-all NITs ticket https://mattermost.atlassian.net/browse/MM-33787

hanzei commented on src/clients/zendesk.ts:
```diff
@@ -1,14 +1,63 @@
+
+export const newZDClient = (context: AppContext): ZDClient => {
+    // get active mattermost user ID
+    const mmUserID = context.acting_user_id || '';
```
Is there a case where `context.acting_user_id` is undefined?

jfrerich replied:
put into catch-all NITs ticket https://mattermost.atlassian.net/browse/MM-33787

jfrerich commented on src/app/app.ts:
```diff
@@ -1,71 +1,138 @@
         const botUserID = configStore.getBotUserID();
-        await tryPromiseWithMessage(adminClient.addToTeam(context.team_id, botUserID), 'Failed to add bot to team');
-        await tryPromiseWithMessage(adminClient.addToChannel(botUserID, context.channel_id), 'Failed to add bot to team');
+        const addToTeamReq = adminClient.addToTeam(this.context.team_id, botUserID);
```
I added a ticket to add the call response to Zendesk after it is implemented. https://mattermost.atlassian.net/browse/MM-33772

hanzei commented on src/utils/bindings.ts:
```diff
@@ -0,0 +1,60 @@
+        location: AppBindingLocations.POST_MENU_ITEM,
+        bindings,
+    };
+    return binding || {};
```
Why is the `||` needed?

jfrerich replied:
put into catch-all NITs ticket https://mattermost.atlassian.net/browse/MM-33787

hanzei reviewed (commented):
A few things caught my eyes

jfrerich commented on src/forms/subscriptions.ts:
```diff
@@ -0,0 +1,401 @@
+        const searchReq = this.zdClient.triggers.search(SubscriptionFields.PrefixTriggersTitle);
+        const triggers = await tryPromiseWithMessage(searchReq, 'Failed to fetch triggers');
+        for (const trigger of triggers) {
+            await this.addChannelTrigger(trigger);
```
ticket created. https://mattermost.atlassian.net/browse/MM-33779

jfrerich commented on src/utils/utils.ts:
```diff
@@ -1,24 +1,55 @@
+
+const getDisplaySubTitleOption = (option: zdSubscriptionFieldOption): string => {
+    const re = new RegExp(SubscriptionFields.RegexTriggerTitle);
+    const newTitle = option.title.match(re)[2];
```
MM discussion link added to this ticket. https://mattermost.atlassian.net/browse/MM-33783

jfrerich commented on src/forms/subscriptions.ts:
```diff
@@ -0,0 +1,401 @@
+// FormFields retrieves viewable modal app fields
+class FormFields extends BaseFormFields {
+    teamTriggers: any
+    teamChannelsWithSubs: Channel[]
```
I wanted to make it known that they are channels scoped to a particular team

jfrerich replied:
@mickmister, what do you think?

jfrerich replied:
put into catch-all NITs ticket https://mattermost.atlassian.net/browse/MM-33787

jfrerich commented on src/forms/subscriptions.ts:
```diff
@@ -0,0 +1,401 @@
+
+    async addChannelTrigger(trigger: any): Promise<void> {
+        const channelID = getChannelIDFromTriggerTitle(trigger.title);
+        const channel = await this.mmClient.getChannel(channelID);
```
put into catch-all NITs ticket https://mattermost.atlassian.net/browse/MM-33787

jfrerich commented on src/restapi/fIncomingWebhooks.ts:
```diff
@@ -0,0 +1,59 @@
+        return `Ticket (${ticketLink}) -- \`${auditEvent.field_name}\` changed from \`${auditEvent.previous_value}\` to \`${auditEvent.value}\``;
+
+    default:
+        return `type not found. type = ${auditEvent.type})`;
```
put into catch-all NITs ticket mattermost.atlassian.net/browse/MM-33787

jfrerich commented on src/restapi/fIncomingWebhooks.ts:
```diff
@@ -0,0 +1,59 @@
+    const auditReq = zdClient.tickets.exportAudit(ticketID);
+    const ticketAudits = await tryPromiseWithMessage(auditReq, `Failed to get ticket audits for ticket ${ticketID}`);
+    const ticketAudit = ticketAudits.slice(-1)[0];
+    const auditEvent = ticketAudit.events[0];
```
put into catch-all NITs ticket mattermost.atlassian.net/browse/MM-33787

jfrerich commented:
@mickmister, I addressed many of your comments and made tickets for the rest. 

@hanzei, I added all of your comments to a ticket -> here https://mattermost.atlassian.net/browse/MM-33787

These will all get addressed, but I want to get this merged and the kvstore so that Dylan can test the latest changes.

Merging now!  Thanks for all the solid feedback!
